File: Programmers/코딩테스트_고득점_Kit/DFS_BFS/level3_아이템줍기.cpp
// C++에서는 조건을 줄 때 3<x<5 이런 식으로 하는 게 안된다.
// 이걸 까먹었어서 몇 시간 날렸다 ^^~
// 항상 같은 로직으로 짰는데 답이 이상하면 언어에 따른 문법 오류가 있는지
// 체크하자..........

#include <string>
#include <vector>

using namespace std;

int map[102][102] = {0};
int visited[102][102] = {0};
int directions[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

int bfs(int character_x, int character_y, int item_x, int item_y) {
  vector<pair<int, int>> q;
  q.push_back(make_pair(character_y, character_x));
  visited[character_y][character_x] = 1;

  while (!q.empty()) {
    int current_y = q.front().first, current_x = q.front().second;
    q.erase(q.begin());  // pop
    for (int i = 0; i < 4; i++) {
      int new_y = current_y + directions[i][0],
          new_x = current_x + directions[i][1];
      if (visited[new_y][new_x] == 0 && map[new_y][new_x] == 1) {
        visited[new_y][new_x] = visited[current_y][current_x] + 1;
        q.push_back(make_pair(new_y, new_x));
      }
      if (new_y == item_y && new_x == item_x) {
        break;
      }
    }
  }
  return (visited[item_y][item_x] - 1) / 2;
}

int solution(vector<vector<int>> rectangle, int characterX, int characterY,
             int itemX, int itemY) {
  for (int i = 0; i < rectangle.size(); i++) {
    int bottom_left_x = rectangle[i][0] * 2,
        bottom_left_y = rectangle[i][1] * 2, top_right_x = rectangle[i][2] * 2,
        top_right_y = rectangle[i][3] * 2;

    for (int y = bottom_left_y; y < top_right_y + 1; y++) {
      for (int x = bottom_left_x; x < top_right_x + 1; x++) {
        if ((bottom_left_x < x && x < top_right_x) &&
            (bottom_left_y < y && y < top_right_y)) {
          map[y][x] = 2;  // 직사각형 내부
        } else if (map[y][x] != 2) {
          map[y][x] = 1;  // boundary
        }
      }
    }
  }

  int answer = bfs(characterX * 2, characterY * 2, itemX * 2, itemY * 2);
  return answer;
}

File: Programmers/코딩테스트_고득점_Kit/DFS_BFS/level3_아이템줍기.py
# 좌표 두 배 해서 5, 5.5, 6..이런식으로 진행될 수 있도록 하기
from collections import deque

map = [[0 for _ in range(102)] for _ in range(102)]
visited = [
    [0 for _ in range(102)] for _ in range(102)
]  # bfs 수행 위함. distance 정보 저장용으로 씀 (메모리 덜 쓰기 위해)
directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]  # 시계방향 검사 위함


def bfs(character_x, character_y, item_x, item_y):
    q = deque([(character_x, character_y)]) 

    visited[character_y][character_x] = 1  # 방문정보 및 거리정보 기록

    while len(q) != 0:  # 큐 빌 때 까지
        current_x, current_y = q.popleft()
        for direction in directions:
            new_y, new_x = (
                current_y + direction[0],
                current_x + direction[1],
            )  # 새로운 좌표 계산
            if (
                not visited[new_y][new_x] and map[new_y][new_x] == 1
            ):  # boundary이면서 아직 방문 안했으면
                visited[new_y][new_x] = visited[current_y][current_x] + 1  # 거리정보 갱신
                q.append((new_x, new_y))
                if new_y == item_y and new_x == item_x:  # 도착지점 도달한 경우
                    break

    return (
        visited[item_y][item_x] - 1
    ) // 2  # 좌표 2개했으며, 첫 시작점 distance를 1로 했었기 때문에 값 고려해 반환


def solution(rectangle, characterX, characterY, itemX, itemY):
    for coordinates in rectangle:  # 주어진 좌표 기반으로 맵 생성
        bottom_left_x, bottom_left_y, top_right_x, top_right_y = (
            coordinates[0] * 2,
            coordinates[1] * 2,
            coordinates[2] * 2,
            coordinates[3] * 2,
        )  # 각 좌표 정보 추출 및 저장

        for y in range(bottom_left_y, top_right_y + 1):
            for x in range(bottom_left_x, top_right_x + 1):
                if bottom_left_x < x < top_right_x and bottom_left_y < y < top_right_y:
                    map[y][x] = 2  # 직사각형 내부
                elif map[y][x] != 2:  # 기존 맵에서 직사각형 내부가 아닌 경우
                    map[y][x] = 1  # boundary

    answer = bfs(characterX * 2, characterY * 2, itemX * 2, itemY * 2)
    return answer
